compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -633,6 +636,15 @@ impl<'a> Parser<'a> {
         self.parse_expr_unary(lo, UnOp::Not)
     }
 
+    /// Parse `box expr` - this syntax has been removed, but we still parse this
+    /// for now to provide an automated way to fix usages of it
+    fn parse_expr_box(&mut self, lo: Span) -> PResult<'a, (Span, ExprKind)> {
+        let (span, expr) = self.parse_expr_prefix_common(lo)?;
+        let code = self.sess.source_map().span_to_snippet(span.with_lo(lo.hi())).unwrap();
+        self.sess.emit_err(errors::BoxSyntaxRemoved { span, code: code.trim() });
+        Ok((span, expr.into_inner().kind))
```

Comment:
I'm not sure this recovery is enough. One might have to add a `Box::new` call around it to support cases like:

```Rust
fn foo(_v: Box<u8>) {}
fn main() {
    foo(box 0);
    let v: Box<()> = box ();
}
```

If the code to add a call to `Box::new` for recovery purposes is too involved, one could think about keeping `ExprKind::Box` around in both HIR and AST until the suggestion is removed from the error in a couple of months (maybe 3 months are enough?).

Reply:
Updated to parse `box x` as `::std::boxed::Box::new(x)`, while emitting the same diagnostic